odilo.py

```diff
@@ -0,0 +1,598 @@
+        return self._do_post(self.library_api_base_url + url, payload, headers, **kwargs)
+
+    @staticmethod
+    def _update_credential(credential, odilo_data):
+        """Copy Odilo OAuth data into a Credential object."""
+        credential.credential = odilo_data['token']
+        expires_in = (odilo_data['expiresIn'] * 0.9)
+        credential.expires = datetime.datetime.utcnow() + datetime.timedelta(seconds=expires_in)
+
+    def get_metadata(self, record_id):
+        identifier = record_id
+        if isinstance(record_id, Identifier):
+            identifier = record_id.identifier
+
+        url = self.RECORD_METADATA_ENDPOINT.format(recordId=identifier)
+
+        status_code, headers, content = self.get(url)
+        if status_code == 200 and content:
+            return content
+        else:
+            msg = 'Cannot retrieve metadata for record: ' + record_id + ' response http ' + status_code
+            if content:
+                msg += ' content: ' + content
+            self.log.warn(msg)
+            return None
+
+    def get_availability(self, record_id):
+        url = self.RECORD_AVAILABILITY_ENDPOINT.format(recordId=record_id)
+        status_code, headers, content = self.get(url)
+        content = json.loads(content)
+
+        if status_code == 200 and len(content) > 0:
+            return content
+        else:
+            msg = 'Cannot retrieve availability for record: ' + record_id + ' response http ' + status_code
+            if content:
+                msg += ' content: ' + content
+            self.log.warn(msg)
+            return None
+
+    @staticmethod
+    def _do_get(url, headers, **kwargs):
+        # More time please
+        if 'timeout' not in kwargs:
+            kwargs['timeout'] = 60
+
+        if 'allow_redirects' not in kwargs:
+            kwargs['allow_redirects'] = True
+
+        response = HTTP.get_with_timeout(url, headers=headers, **kwargs)
+        return response.status_code, response.headers, response.content
+
+    @staticmethod
+    def _do_post(url, payload, headers, **kwargs):
+        # More time please
+        if 'timeout' not in kwargs:
+            kwargs['timeout'] = 60
+
+        return HTTP.post_with_timeout(url, payload, headers=headers, **kwargs)
+
+
+class MockOdiloAPI(OdiloAPI):
+    @classmethod
+    def mock_collection(cls, _db):
+        library = DatabaseTest.make_default_library(_db)
+        collection, ignore = get_one_or_create(
+            _db, Collection,
+            name="Test Odilo Collection",
+            create_method_kwargs=dict(
+                external_account_id=u'library_id_123',
+            )
+        )
+        integration = collection.create_external_integration(
+            protocol=ExternalIntegration.ODILO
+        )
+        integration.username = u'username'
+        integration.password = u'password'
+        integration.setting(OdiloAPI.LIBRARY_API_BASE_URL).value = u'http://localhost:8080/api/v2'
```

Comment:
This should be a completely fake URL that will never work, to ensure that tests never actually try to make an HTTP request.

Reply:
ok, changed !!